pkg/tbtc/deposit_sweep.go

```diff
@@ -4,15 +4,227 @@ import (
+	sweepingWallet  wallet
+	signingExecutor *signingExecutor
+	proposal        *DepositSweepProposal
+}
+
+func newDepositSweepAction(
+	chain Chain,
+	btcChain bitcoin.Chain,
+	sweepingWallet wallet,
+	signingExecutor *signingExecutor,
+	proposal *DepositSweepProposal,
+) *depositSweepAction {
+	return &depositSweepAction{
+		chain:           chain,
+		btcChain:        btcChain,
+		sweepingWallet:  sweepingWallet,
+		signingExecutor: signingExecutor,
+		proposal:        proposal,
+	}
+}
+
+// TODO: Cover this function with unit tests once everything is completed.
+func (dsa *depositSweepAction) execute() error {
+	walletPublicKeyBytes, err := marshalPublicKey(dsa.wallet().publicKey)
+	if err != nil {
+		return fmt.Errorf("cannot marshal wallet public key: [%v]", err)
+	}
+
+	actionLogger := logger.With(
+		zap.String("wallet", fmt.Sprintf("0x%x", walletPublicKeyBytes)),
+		zap.String("action", dsa.actionType().String()),
+	)
+
+	depositExtraInfo := make(
+		[]struct {
+			*Deposit
+			FundingTx *bitcoin.Transaction
+		},
+		len(dsa.proposal.DepositsKeys),
+	)
+
+	depositsCount := len(dsa.proposal.DepositsKeys)
+
+	actionLogger.Infof("gathering prerequisties for proposal validation")
+
+	for i, depositKey := range dsa.proposal.DepositsKeys {
+		depositDisplayIndex := fmt.Sprintf("%v/%v", i+1, depositsCount)
+
+		depositLogger := actionLogger.With(
+			zap.String(
+				"depositFundingTxHash",
+				fmt.Sprintf("0x%x", depositKey.FundingTxHash),
+			),
+			zap.String("depositIndex", depositDisplayIndex),
+		)
+
+		depositLogger.Infof("checking confirmations count for deposit's funding tx")
+
+		confirmations, err := dsa.btcChain.GetTransactionConfirmations(
+			depositKey.FundingTxHash,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"cannot get funding tx confirmations count "+
+					"for deposit [%v]: [%v]",
+				depositDisplayIndex,
+				err,
+			)
+		}
+
+		if confirmations < depositSweepRequiredFundingTxConfirmations {
+			return fmt.Errorf(
+				"funding tx of deposit [%v] has only [%v/%v] of "+
+					"required confirmations",
+				depositDisplayIndex,
+				confirmations,
+				depositSweepRequiredFundingTxConfirmations,
+			)
+		}
+
+		depositLogger.Infof("fetching deposit's extra data")
+
+		fundingTx, err := dsa.btcChain.GetTransaction(depositKey.FundingTxHash)
+		if err != nil {
+			return fmt.Errorf(
+				"cannot get funding tx data for deposit [%v]: [%v]",
+				depositDisplayIndex,
+				err,
+			)
+		}
+
+		depositRequest, err := dsa.chain.GetDepositRequest(
+			depositKey.FundingTxHash,
+			depositKey.FundingOutputIndex,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"cannot get on-chain request data for deposit [%v]: [%v]",
+				depositDisplayIndex,
+				err,
+			)
+		}
+
+		// Worth mentioning this should be treated as an estimation, giving
+		// the caveats mentioned in the docstring of GetBlockNumberByTimestamp.
+		revealBlock, err := dsa.chain.GetBlockNumberByTimestamp(
+			uint64(depositRequest.RevealedAt.Unix()),
+		)
```

Comment:
The logic in `GetBlockNumberByTimestamp` is quite complicated and I am not sure if we really need it.

Both depositor address and wallet public key hash are indexed fields and we can easily filter by them.
How many deposits to the given wallet a single depositor address can make at maximum? 1000? 5000? 10000? 

If we are afraid we can exceed the limits of popular Ethereum node providers when pulling events (Infura, Alchemy, Quicknode), we should implement partial event pulls like in https://github.com/keep-network/tbtc-v2/pull/604 in the smart contract bindings if protection about this kind of situations is not already there.

Reply:
The general logic of how `eth_getLogs` work under the hood is (see https://github.com/ethereum/go-ethereum/blob/master/eth/filters/filter.go#L97):
- Receive a block range. If not provided use [1, latest]
- Iterate over block headers in the range and check the bloom filter
- If there are logs matching the filter criteria, return them

In other words, we must specify a good range to make the call efficient. The indexed field is just used to limit the number of returned data. Both the range and amount of returned data are important in the context of node provider limitations.

Partial event pulling is not a solution in this context. We still need a starting block range to fetch that must be established. Otherwise, we will partially pull events from the range [1, latest].  Partial pulls help when we have a limited range that is too wide or the amount of returned logs are too big. However, we still need to establish a range first and this is why we have `GetBlockNumberByTimestamp`. Also, in the case of a deposit sweep, time is crucial. We must gather the necessary data as quickly as possible and I think the presented solution is all we can do here.

Reply:
Let's take a step back to make sure we are on the same page.

Those are the limits of popular Ethereum node providers I could find:

> Infura has a cap on requests of 10,000 events per query. 

https://blog.infura.io/post/ethereum-rpcs-methods#using-eth-get-logs

> You can make eth_getLogs requests with up to a 2K block range and 150MB limit on the response size OR you can request any block range with a cap of 10K logs in the response.

https://docs.alchemy.com/docs/deep-dive-into-eth_getlogs#making-a-request-to-eth_getlogs

> eth_getLogs and eth_newFilter are limited to a 10,000 blocks range

https://support.quicknode.com/hc/en-us/articles/10258449939473-Is-there-a-block-range-limit-for-querying-logs-and-events-

The first problem we have are API limits. Assuming a 12s average block time, Ethereum will produce 7200 blocks/24h. Even if we use `GetBlockNumberByTimestamp`, sweeping deposits older than two days with Quicknode will not be possible unless the underlying Ethereum communication layer does something smart.

For Alchemy and Infura, realistically, it doesn't look like a problem because it's rather unrealistic a single depositor can make 10k deposits to the same wallet.

Nevertheless, I propose we confirm where we are by executing a test and/or checking the code. If there is such a limitation, we should add a GitHub issue and prioritize it accordingly. 

The second problem we have is how efficient `GetBlockNumberByTimestamp` actually is compared to e.g. checking always the last N blocks.

I did some experiments and it takes:
- 5 `blockByNumber` calls to evaluate block number for yesterday timestamp `1681997871`
- 7 `blockByNumber` calls to evaluate block number for week ago timestamp `1681479471`
- 8 `blockByNumber` calls to evaluate block number for 6 months ago timestamp `1666263335`
- The function executes `blockByNumber` calls for 15 seconds until it times out for 12 months ago timestamp `1650538535`.

The last experiment is concerning because I do not know if that's a bug in `GetBlockNumberByTimestamp` or if it just takes that long because it was 12 months ago. Quite surprised though given it takes 8 blocks for a timestamp 6 months ago.

For this conversation, let's assume the depositor is not doing more than 10,000 deposits to the given wallet. 
Let's also assume the wallet is sweeping deposits not older than a week. This is how it should work once we activate sweeping and deal with pre-sweeping wallets. 

If we hardcode N to 100,000 blocks (~ a week) then it should take:
- One Infura call
- One Alchemy call
- Ten Quicknode calls

This is less than executing additional 7 blocks to find the block timestamp. 

It could be worth checking how long it takes to execute such a call on Infura or Alchemy but I doubt it will take more than 7 additional calls to `blockByNumber`.

The situation we are currently in is exceptional because we'll have to sweep deposits from 6 months ago. But I think it's still all doable with one call to Infura and one call to Alchemy setting the start block to the time when Bridge contract was deployed. 

My tl;dr; is:
- I am concerned it looks that  `GetBlockNumberByTimestamp` does not always work/does not always work efficient enough.
- We should see how long it takes to scan 6-12 months of blocks on Ethereum node and then take the decision if 
`GetBlockNumberByTimestamp` is an optimization or not.
- Quicknode needs special support no matter if we have `GetBlockNumberByTimestamp` or not.

Reply:
> The first problem we have are API limits. Assuming a 12s average block time, Ethereum will produce 7200 blocks/24h. Even if we use GetBlockNumberByTimestamp, sweeping deposits older than two days with Quicknode will not be possible unless the underlying Ethereum communication layer does something smart.

Not sure if I understand correctly but Quicknode has a general limitation "no more than 10k requested blocks" not necessarily from `[now-10k, now]` range. In this case `GetBlockNumberByTimestamp` allows us to limit the requested range to the minimum.

> For Alchemy and Infura, realistically, it doesn't look like a problem because it's rather unrealistic a single depositor can make 10k deposits to the same wallet.

This is not the problem about the number of returned data as this, as you noticed, is limited using the depositor and wallet indexed fields. The problem is about the requested block range. The wider the block range the more time the node needs to fetch the logs. Even if the given provider doesn't have specific limitations on the requested range, the time of a single request can be problematic. There may also be some timeouts on the provider side that may bite us. 

Also, we should also consider self-hosted Ethereum nodes that can be easily nuked by a wide block range `eth_getLogs` call if such a node does not have appropriate protections. See Alechemy's note about `eth_getLogs` vulnerabilities: https://docs.alchemy.com/docs/deep-dive-into-eth_getlogs#making-a-request-to-eth_getlogs

Regarding the performance of `GetBlockNumberByTimestamp`: its efficiency is not linearly correlated with how the requested timestamp is far from now. The clue is the actual average block time between the predicted and current block. Since the Ethereum merge in September 2022, the average block time is slightly about 12 seconds. The `GetBlockNumberByTimestamp` assumes a bit higher average block time (13 seconds) which allows it to make fewer jumps to the requested block and not overshoot it. Here is why the number of calls for 6 months ago timestamp is quite low. This timestamp is after September 2022 and the actual average block time has a stable value above 12 seconds. On the other hand, the problematic 1-year-ago timestamp from your exeriment is before September 2022 and the actual average block time between that timestamp and now is much more variable and much bigger than 12 seconds. That means `GetBlockNumberByTimestamp` will jump by far more blocks to the past than necessary. In that case, the function will start going forward block by block to compensate for the error and will need much more `blockByNumber` requests. The timeout you observed comes from `blockByNumber` and is set to 10 seconds initially. This timeout may be too short for a node that must get many blocks from the far past. Here is what you observed in the last experiment. The conclusion is that this is not a bug directly in `GetBlockNumberByTimestamp` but we should rather think about using a longer timeout.

Reply:
> > The first problem we have are API limits. Assuming a 12s average block time, Ethereum will produce 7200 blocks/24h. Even if we use GetBlockNumberByTimestamp, sweeping deposits older than two days with Quicknode will not be possible unless the underlying Ethereum communication layer does something smart.
> 
> Not sure if I understand correctly but Quicknode has a general limitation "no more than 10k requested blocks" not necessarily from [now-10k, now] range. In this case GetBlockNumberByTimestamp allows us to limit the requested range to the minimum.

Yes, but I think this is still not enough to support Quicknode. Say that `GetBlockNumberByTimestamp` returned block `N - 20000` where `N` is the current block. If we do a request with the `[N - 20000, N]` range, it will fail for Quicknode. We need two separate requests: `[N-20000, N - 10000]` and `[N-10000, N]`. I do not think it is something we need to worry about in this PR but we should address it soon after. It could also be addressed by making it clear in the docs that Quicknode is currently not supported and opening an issue for implementing a similar mechanism as in https://github.com/keep-network/tbtc-v2/pull/604 for Go. But let's first confirm this is really the case and nothing in our Go bindings or `go-ethereum` makes it work. Just a simple call to Quicknode for past events with more than 10,000 blocks in range should do the trick.

Regarding `GetBlockNumberByTimestamp` in general, what I am trying to say is that I am not sure if this is more performant than just selecting the range of the last month's blocks. For every deposit being swept, we will have to execute 7-8 additional calls. Maybe that takes more time than just one call with `[N - 30 * 7200, N]` blocks. We can not do it now given we will be sweeping deposits from the last few months though.

So tl;dr; of my proposition would be:
1. Leave this as-is.
2. Once this PR is merged, please check if this works with Quicknode. If not, let's add a note to the docs that Quicknode is not supported.

Separately, I'll open an issue to investigate the performance of `GetBlockNumberByTimestamp` compared to just making a call with a wider block range.

How does it sound?

Reply:
> Yes, but I think this is still not enough to support Quicknode. Say that GetBlockNumberByTimestamp returned block N - 20000 where N is the current block. If we do a request with the [N - 20000, N] range, it will fail for Quicknode. We need two separate requests: [N-20000, N - 10000] and [N-10000, N]. 

Thanks to `GetBlockNumberByTimestamp` we don't have to use such a wide range. We are actually using just a small margin (10 blocks) on both sides of the estimated reveal block to handle possible corner cases. See https://github.com/keep-network/keep-core/blob/f49238d8ca102eefc7b9b9bb6aaab0e0d14622d4/pkg/tbtc/deposit_sweep.go#L157

That means Quicknode will work as we use only 20-block ranges which are far from the 10k blocks limit.

Regarding your proposition, it sounds good! We should explore the performance of `GetBlockNumberByTimestamp` later and maybe introduce some kind of event indexer that would cache chain events and will be used as the first-choice source of data and only falling back to the current mechanism with `GetBlockNumberByTimestamp` in case of a cache miss.